### DISCUSSION TITLE
Please help with Keil

### DISCUSSION THREAD

**Discussion by vorosj** [hathach/tinyusb#1561] [General]
I have hard time porting an example to Keil ide. Could you please share an example project, that's compiled in Keil 5 and working? I am working with an STM32F103, but of course any STM32 project helps.

I am not familiar with makefiles.

If I follow the steps according to "getting started", Keil includes only the 3 .c files in tinyusb/source, but no others from the map structure. If I add manually the missing files (there are many), not sure if I don't include unnecessary files and get errors etc. There must be a more practical way.

**Comment by pigrew** [comment]
(I don't have experience with Keil, but I've seen others use it...)

If I were you, I'd setup the Kiel project "manually", like you imply... 

Include the tinyusb "src" directory in your project, but exclude everything in src/portable except for st/fsdev. You also need to include the relevant BSP sources and example sources.

The examples use ST HAL, but I'm assuming you are using CMSIS drivers? If so, you'll need to rewrite the BSP  to use those drivers (unless you do actually want to use ST HAL).

There are a few DEFINE you need to set to declare which MCU you are using, but I think they likely will already be set by Keil.

**Reply by vorosj** [reply]
Thank you for help.
I try it hard, but no succes yet.

The first problem, that there is no Keil compiler option in the tusb_compiler.h. I am trying to select gcc in the tusb_compiler.h, and add --gnu option in the Keil settings, but not sure it works. Many errors.

I think first I use the HAL device driver, enable the USB device interrups, no HAL middleware, and replace the ST irq  service routines with the tusb one. This part seams easy.

The various settings, macros etc. is the hard part. That's why any working Keil project could help a lot.

Another problem, that I can't make the example files in windows11, the path system is not compatible (it's a known tinyusb makefile bug, but can't find a clear solution). I makes really hard to getting started.

**Comment by hathach** [comment]
> Could you please share an example project,

Unfortunately I don't have any to share. Keil is windows only, cost ton of money, and take lots of effort to maintain considering the amount of examples and ports supported by this repo. It is just much faster to use makefile. 

For running make with windows, try out #656 suggestion from other user to see if that helps

> The various settings, macros etc. is the hard part. That's why any working Keil project could help a lot.

I don't see why this is a hard part, they are plain text in the makefile make.mk rule.mk. Just migrate them to keil option. Maybe if you be more specific on the issue, other keil users may help you with. And yes, you need to enable gnu compatible mode in keil

**Reply by vorosj** [reply]
Hi, thanks for reply.
I am not familiar with Linux, command lines, makefiles etc.
Tried to make an example project (without really undersatnding what I do), and get error messages.
```
c:\0work\keil_projects\tinyUSB\tinyusb-master>make -C examples/device/cdc_msc BOARD=stm32f103_bluepill
make: Entering directory 'c:/0work/keil_projects/tinyUSB/tinyusb-master/examples/device/cdc_msc'
CC dcd_stm32_fsdev.o
arm-none-eabi-gcc: error: unrecognized command line option '-Wcast-function-type'
arm-none-eabi-gcc: error: unrecognized command line option '-Wnull-dereference'
make: *** [../../rules.mk:131: _build/stm32f103_bluepill/obj/src/portable/st/stm32_fsdev/dcd_stm32_fsdev.o] Error 1
make: Leaving directory 'c:/0work/keil_projects/tinyUSB/tinyusb-master/examples/device/cdc_msc'
```
I allways worked in IDE's, don't have to care about these command line parameters etc. For me these are parts under hood, not things a normal driver has to care about.

Any of the example projects ported to Keil (or STM32CubeIDE, Embitz etc), that is working, could help a lot.  That's a starting point, I can modify, check what files are involded, what are the defines etc. After that I can use it on other projects too.

These makefiles, Linux stuff etc like chinese text without google translation.

**Reply by pigrew** [reply]
Quite a while ago, I did use TrueStudio, in a STM32CubeMX project... I didn't reference the makefiles at all... Just added the TinyUSB sources to the project.

You seen to be using gcc and not armclang? I thought Keil normally used armclang? Which version of gcc are you using?

Those particular warning flags were added in gcc 6.1 (or so...), so perhaps they could be simply removed from the makefiles. I'm not sure which version of the C standard, and which versions of gcc are officially supported by TinyUSB.

**Reply by hathach** [reply]
makefile may look like alien language to you, but it is common language in OSS. If you only want to work with IDE (specially one that cost an arm and a leg) without spending your time to do the conversion. You should stick to the vendor/manufacturer package/examples. Since they have the time + human resource to support these.

**Comment by vorosj** [comment]
Now I tried the win11 builtin Ubuntu to compile the examples. I am not familiar with Linux at all.
Following the getting started instructions, everything went well, git cloned the library, installed make etc. until
```
mmormota@DESKTOP-S2K747J:~/tinyusb$ make BOARD=stm32f103_bluepill get-deps
make: *** No rule to make target 'get-deps'.  Stop.
```
I suppose it has to install the f103 core and peripheral libraries? What is the problem?

**Reply by hathach** [reply]
you need to run make in one of the example folder. I will update the docs to make it clearer.

**Comment by vorosj** [comment]
AFAIK Keil uses ARMCC (same as ARMCLANG?).
I just installed a new arm-none-eabi-gcc (GNU Toolchain for the Arm Architecture 11.2-2022.02 (arm-11.16)) 11.2.1 20220111

**Reply by vorosj** [reply]
Thank you, now I can build it in the wsl Linux window.

**Comment by vorosj** [comment]
I ported it to Embitz. 
The uac2_headset example is not working on the stm32f103. After a fast debugging effort I suspect isochronous transfer is not supported on this family?

**Reply by pigrew** [reply]
Correct, see #1249
